Data/ByteString/Internal/Type.hs

```diff
@@ -229,6 +231,38 @@ pokeFpByteOff :: Storable a => ForeignPtr b -> Int -> a -> IO ()
 pokeFpByteOff fp off val = unsafeWithForeignPtr fp $ \p ->
   pokeByteOff p off val
 
+-- | Most operations on a 'ByteString' need to read from the buffer
+-- given by its @ForeignPtr Word8@ field.  But since most operations
+-- on @ByteString@ are (nominally) pure, their implementations cannot
+-- see the IO state thread that was used to initialize the contents of
+-- that buffer.  This means that under some circumstances, these
+-- buffer-reads may be executed before the writes used to initialize
+-- the buffer are executed, with unpredictable results.
+--
+-- 'deferForeignPtrAvailability' exists to help solve this problem.
+-- At runtime, a call @'deferForeignPtrAvailability' x@ is equivalent
+-- to @'pure' x@, but the former is more opaque to the simplifier, so
+-- that reads from the pointer in its result cannot be executed until
+-- the (no-op) @'deferForeignPtrAvailability' x@ call is complete.
+deferForeignPtrAvailability :: ForeignPtr a -> IO (ForeignPtr a)
+deferForeignPtrAvailability (ForeignPtr addr0# guts) = IO $ \s0 ->
+  -- The "lazy runRW#" gunk gets erased during CorePrep,
+  -- leaving just the unboxed tuple, so this should have no direct overhead.
+  case lazy runRW# (\_ -> (# s0, addr0# #)) of
+    (# s1, addr1# #) -> (# s1, ForeignPtr addr1# guts #)
+
+unsafeDupablePerformIO :: IO a -> a
+-- Why does this exist? As of base-4.18.0.0, the version of
+-- unsafeDupablePerformIO in base prevents unboxing of its results
+-- with an opaque call to lazy, for reasons described in
+-- Note [unsafePerformIO and strictness].  Even if we accept the
```

Comment:
This can use a link to https://hackage.haskell.org/package/base-4.18.0.0/docs/src/GHC.IO.Unsafe.html#line-30.